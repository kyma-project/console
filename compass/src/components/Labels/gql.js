import gql from 'graphql-tag';

export const CREATE_LABEL = gql`
  mutation createLabelDefinition($in: LabelDefinitionInput!) {
    createLabelDefinition(in: $in) {
      key
      schema
    }
  }
`;
