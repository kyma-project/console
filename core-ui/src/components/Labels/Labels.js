import React from 'react';
import PropTypes from 'prop-types';
import { FormInput } from 'fundamental-react';

LabelsInput.propTypes = {
  labels: PropTypes.object,
  updateLabels: PropTypes.func,
  updateErrorMessage: PropTypes.func,
};

export default function LabelsInput({
  labels,
  updateLabels,
  updateErrorMessage,
}) {
  function validateLabel(event) {
    const newLabel = event.target.value.trim();
    if (newLabel === '') return;
    const separator = '=';
    if (!(newLabel.split(separator).length === 2)) {
      updateErrorMessage(
        `Invalid label '${newLabel}'. A key and value should be separated by a '='`,
      );
      return;
    }

    const regex = /([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]/;
    const key = newLabel.split(separator)[0];
    const value = newLabel.split(separator)[1];
    const foundKey = key.match(regex);
    const foundVal = value.match(regex);
    const isKeyValid = foundKey && foundKey[0] === key && key !== '';
    const isValueValid = (foundVal && foundVal[0] === value) || value === '';
    if (!isKeyValid || !isValueValid) {
      updateErrorMessage(
        `Invalid label '${key}${separator}${value}'. In a valid label, a key cannot be empty, a key/value consists of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character.`,
      );
      return;
    }
    if (labels[key]) {
      updateErrorMessage(`Label with key value '${key}' already exists.`);
      return;
    }

    const newLabels = { ...labels };
    newLabels[key] = value;
    updateLabels(newLabels);
    updateErrorMessage('');
    event.target.value = '';
  }

  return (
    <FormInput
      id="labels"
      type="text"
      placeholder="Add labels"
      autoComplete="off"
      onKeyDown={e => {
        if (e.key === 'Enter' || e.key === ' ') {
          validateLabel(e);
        }
      }}
      onBlur={e => {
        validateLabel(e);
      }}
    />
  );
}
