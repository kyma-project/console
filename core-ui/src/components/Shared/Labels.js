import React from 'react';
import PropTypes from 'prop-types';
import {
  FormItem,
  FormLabel,
  FormInput,
  Token,
  FormMessage,
  InlineHelp,
} from 'fundamental-react';

LabelsInput.propTypes = {
  labels: PropTypes.object,
  updateLabels: PropTypes.func,
};

export default function LabelsInput({ labels, updateLabels }) {
  const [errorMessage, setErrorMessage] = React.useState('');

  function addLabel(event) {
    event.preventDefault();
    setErrorMessage('');

    const newLabel = event.target.value.trim();
    if (!(newLabel.split('=').length === 2)) {
      return;
    }

    const regex = /(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])=(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]|$)/;
    const matchingLabel = newLabel.match(regex);
    const isLabelValid =
      matchingLabel && matchingLabel[0] === matchingLabel.input;
    if (!isLabelValid) {
      return;
    }

    const key = newLabel.split('=')[0];
    const value = newLabel.split('=')[1];
    if (labels[key]) {
      setErrorMessage(`Label with key '${key}' already exists.`);
      return;
    }

    const newLabels = { ...labels };
    newLabels[key] = value;
    updateLabels(newLabels);
    event.target.value = '';
  }

  function createLabelsToDisplay() {
    const labelsArray = [];
    for (const key in labels) {
      const value = labels[key];
      const labelToDisplay = `${key}=${value}`;
      labelsArray.push(labelToDisplay);
    }
    return labelsArray;
  }

  function deleteLabel(event) {
    const labelToDisplay = event.target.innerHTML;
    const key = labelToDisplay.split('=')[0];
    const newLabels = { ...labels };
    delete newLabels[key];
    updateLabels(newLabels);
  }

  return (
    <FormItem key="labels">
      <FormLabel htmlFor="labels">
        Labels
        <InlineHelp
          placement="bottom-right"
          text="A key and value should be separated by a '=', a key cannot be empty, a key/value consists of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character."
        />
      </FormLabel>

      {createLabelsToDisplay().map(label => (
        <Token onClick={e => deleteLabel(e)}>{label}</Token>
      ))}

      <FormInput
        id="labels"
        type="text"
        placeholder="Add new label"
        autoComplete="off"
        pattern="^(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])=(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]|$)"
        onKeyDown={e => {
          if (e.key === 'Enter' || e.key === ' ') {
            addLabel(e);
          }
        }}
        onBlur={e => {
          addLabel(e);
        }}
      />

      {errorMessage && <FormMessage type="error">{errorMessage}</FormMessage>}
    </FormItem>
  );
}
