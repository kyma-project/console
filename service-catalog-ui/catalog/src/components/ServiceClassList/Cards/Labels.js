import React from 'react';
import { Label, Tooltip, PanelFooter } from '@kyma-project/react-components';
import './Labels.scss';
import { isStringValueEqualToTrue } from '../../../commons/helpers';
const labelsDescription = {
  'connected-app':
    'This Service Class is connected to a given application by Application Connector.',
  showcase:
    'This Service Class presents a specific functionality. Do not use it on the production cluster.',
};
export function Labels({ labels }) {
  return (
    <PanelFooter className="service-list--card__footer">
      {labels &&
        Object.keys(labels).length > 0 &&
        Object.keys(labels).map(label => {
          if (label === 'local' || label === 'provisionOnlyOnce') {
            return null;
          }
          if (
            (label === 'showcase' &&
              !isStringValueEqualToTrue(labels[label])) ||
            (label === 'connected-app' && !labels[label])
          ) {
            return null;
          }

          return (
            <div className="service-list--card__footer__labels" key={label}>
              <Tooltip content={labelsDescription[label]}>
                <Label cursorType="help">
                  {label === 'connected-app' ? labels['connected-app'] : label}
                </Label>
              </Tooltip>
            </div>
          );
        })}
    </PanelFooter>
  );
}
