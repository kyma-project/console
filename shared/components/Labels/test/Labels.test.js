import renderer from 'react-test-renderer';
import { Labels } from '../Labels';
import React from 'react';

describe('Labels', () => {
  it('Labels should render with no labels', () => {
    const component = renderer.create(<Labels />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Labels should render with labels', () => {
    const component = renderer.create(
      <Labels labels={{ testLabel: 'testValue' }} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
