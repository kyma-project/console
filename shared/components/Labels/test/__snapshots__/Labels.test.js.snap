// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Labels Labels should render with labels 1`] = `
<span
  className="fd-token y-fd-token y-fd-token--no-button y-fd-token--gap"
  role="button"
>
  testLabel=testValue
</span>
`;

exports[`Labels Labels should render with no labels 1`] = `
<span>
  -
</span>
`;
